src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java

```diff
@@ -150,6 +164,49 @@ public void buildVariantsUpdateActions_doesNotRemoveMaster() throws Exception {
     }
 
     @Test
+    public void buildVariantsUpdateActions_withEmptyOldMasterVariantKey() throws Exception {
+        assertMissingMasterVariantKey(OLD_PROD_WITHOUT_MV_KEY_SKU, NEW_PROD_DRAFT_WITH_VARIANTS_MOVE_MASTER,
+            BLANK_OLD_MASTER_VARIANT_KEY);
+    }
+
+    @Test
+    public void buildVariantsUpdateActions_withEmptyNewMasterVariantOrKey() throws Exception {
+        assertMissingMasterVariantKey(OLD_PROD_WITH_VARIANTS, NEW_PROD_DRAFT_WITHOUT_MV, BLANK_NEW_MASTER_VARIANT_KEY);
+        assertMissingMasterVariantKey(OLD_PROD_WITH_VARIANTS, NEW_PROD_DRAFT_WITHOUT_MV_KEY,
+            BLANK_NEW_MASTER_VARIANT_KEY);
+    }
+
+    @Test
+    public void buildVariantsUpdateActions_withEmptyBothMasterVariantKey() throws Exception {
+        assertMissingMasterVariantKey(OLD_PROD_WITHOUT_MV_KEY_SKU, NEW_PROD_DRAFT_WITHOUT_MV_KEY,
+            BLANK_OLD_MASTER_VARIANT_KEY, BLANK_NEW_MASTER_VARIANT_KEY);
+    }
+
+    private void assertMissingMasterVariantKey(final String oldProduct,
+                                               final String newProduct,
+                                               final String ... errorMessages) {
+        final Product productOld = createProductFromJson(oldProduct);
+        final ProductDraft productDraftNew = createProductDraftFromJson(newProduct);
+
+        final List<String> errorsCatcher = new ArrayList<>();
+        final ProductSyncOptions syncOptions = ProductSyncOptionsBuilder.of(mock(SphereClient.class))
+            .setErrorCallBack((message, exception) -> errorsCatcher.add(message))
+            .build();
+
+        final List<UpdateAction<Product>> updateActions =
+            buildVariantsUpdateActions(productOld, productDraftNew, syncOptions, emptyMap());
+        assertThat(updateActions).isEmpty();
+        assertThat(errorsCatcher).hasSize(errorMessages.length);
+
+        // verify all expected error messages
+        for (int i = 0; i < errorMessages.length; i++) {
+            assertThat(errorsCatcher.get(i))
+                .contains(format("Failed to build update actions on the product with key '%s'", productOld.getKey()))
+                .contains(errorMessages[i]);
```

Comment:
why not? :
````java
assertThat(errorsCatcher.get(i)).isEqualTo(
    format("Failed to build update actions on the product with key '%s'. Reason: %",
    productOld.getKey(), errorMessages[i]));
````

Reply:
i just wanted to have some freedom for the error message, so the error message should not be exactly like this, but contain some keywords in any order.
For example, when leading/trailing/redundant spaces/periods/commas are changed in the message, the test is still valid.

I'd say usually i make it even less precise, so smth like:
```java
 assertThat(errorsCatcher.get(i))
    .containsIgnoringCase("failed")
    .contains(productOld.getKey())
    .containsIgnoringCase(errorMessages[i]);
```

In this case changes like in [the comment above](https://github.com/commercetools/commercetools-sync-java/pull/140#discussion_r143719118) do not require test changes.

Reply:
but why is it better to make it flexible for change (for us devs) against having a strict validation of the actual error message? for example, if the logic fails and another error message different than the one that should be actually is triggered, with also the word "failed" could pass the test when it shouldn't.

Reply:
like this: a601aa6a

Reply:
but i verify exact error messages `BLANK_OLD_MASTER_VARIANT_KEY`, `BLANK_NEW_MASTER_VARIANT_KEY`, `BLANK_NEW_MASTER_VARIANT_SKU` and so on